Implement list.index

aisk opened the pull request:
~~Sorry, forgot the lock, plz don't review now.~~
Ready for review.

trotterdylan commented on runtime/list.go:
```diff
@@ -306,6 +306,70 @@ func listNE(f *Frame, v, w *Object) (*Object, *BaseException) {
 }
 
+func listIndex(f *Frame, args Args, _ KWArgs) (*Object, *BaseException) {
+	expectedTypes := []*Type{ListType, ObjectType, IntType, IntType}
```
I think that for the last two parameters, index() accepts any object that implements `__index__`. See for example [strIndex()](https://github.com/google/grumpy/blob/master/runtime/str.go#L238) which calls IndexInt() to pull out an integer index from an arbitrary Object.

trotterdylan commented on runtime/list.go:
```diff
@@ -306,6 +306,70 @@ func listNE(f *Frame, v, w *Object) (*Object, *BaseException) {
+	index := 0
+	found := false
+	var raised *BaseException
+	for i := start; i < stop; i++ {
```
This seems very similar to the implementation of remove() so maybe it could be factored out into a helper function in seq.go that takes a slice and returns the index.

aisk commented:
WIP.

aisk commented:
Updated, but don't refactor the common codes in list.index / list.remove.

I modified the seqFindFisrt, and let it returns the first position where it found the element, but got a deadlock, like which we found in other PR.

Should we add a new function, and don't convert it to a seq, or just keep them just like now?

trotterdylan commented:
Right, I was thinking that you'd need a new function in seq.go, something like:

```
// seqFindElem returns the index of the element in elems that is equal to o, or -1 if no element is equal.
seqFindElem(f *Frame, elems []*Object, o *Object) (int, *BaseException)
```

aisk commented:
Updated.

trotterdylan commented on runtime/seq.go:
```diff
@@ -156,6 +156,7 @@ func seqFindFirst(f *Frame, iterable *Object, pred func(*Object) (bool, *BaseExc
 	if raised != nil {
 		return false, raised
 	}
+	index := 0
```
I think this is leftover from some obsolete changes?

trotterdylan commented on runtime/list.go:
```diff
@@ -338,6 +338,72 @@ func listNE(f *Frame, v, w *Object) (*Object, *BaseException) {
+		stop = len(l.elems)
+	case 3:
+		numElems := len(l.elems)
+		start, raised = IndexInt(f, args[2])
```
I think you can eliminate a bit of code duplication if you rearrange this as repeated if statements rather than a switch:

```
start, stop := 0, len(l.elems)
if argc >= 2 {
  start, raised = IndexInt(f, args[2])
  ...
}
if argc >= 3 {
  stop, raised = IndexInt(f, args[3])
  ...
}
```

aisk commented:
Updated.

trotterdylan commented on runtime/seq.go:
```diff
@@ -173,6 +173,24 @@ func seqFindFirst(f *Frame, iterable *Object, pred func(*Object) (bool, *BaseExc
 	return false, nil
 }
 
+func seqFindElem(f *Frame, elems []*Object, o *Object, start, stop int) (int, bool, *BaseException) {
```
Rather than passing start and stop, I would have the caller slice the elems: `seqFindElems(f, elems[start:stop], o)`

trotterdylan replied:
Have the return value be `(int, *BaseException)` and have the first return value be -1 when not found.

trotterdylan commented on runtime/list_test.go:
```diff
@@ -91,6 +91,35 @@ func TestListCount(t *testing.T) {
+		{args: wrapArgs(newTestList(0, 1, 2, 3, 4), 3, 0, 999), want: NewInt(3).ToObject()},
+		{args: wrapArgs(newTestList(0, 1, 2, 3, 4), "foo", 0, 999), wantExc: mustCreateException(ValueErrorType, "'foo' is not in list")},
+		{args: wrapArgs(newTestList(0, 1, 2, 3, 4), 3, 999), wantExc: mustCreateException(ValueErrorType, "3 is not in list")},
+		{args: wrapArgs(newTestList(0, 1, 2, 3, 4), 3, 5, 0), wantExc: mustCreateException(ValueErrorType, "3 is not in list")},
```
Can you add some test cases with repeated values and also some values that aren't integers?

trotterdylan replied:
Ping, can you address this request?

aisk commented:
Ready for review again.

trotterdylan commented on runtime/seq.go:
```diff
@@ -173,6 +173,25 @@ func seqFindFirst(f *Frame, iterable *Object, pred func(*Object) (bool, *BaseExc
 }
 
+func seqFindElem(f *Frame, elems []*Object, o *Object) (int, *BaseException) {
+	// for i := start; i < stop; i++ {
```
remove this commented out code

trotterdylan commented on runtime/seq.go:
```diff
@@ -173,6 +173,25 @@ func seqFindFirst(f *Frame, iterable *Object, pred func(*Object) (bool, *BaseExc
+		var raised *BaseException
+		var eq *Object
+		var found bool
+		if eq, raised = Eq(f, elem, o); raised != nil {
```
Slightly less verbose:

```
eq, raised := Eq(f, elem, o)
if raised != nil {
  return -1, raised
}
found, raised := IsTrue(f, eq)
if raised != nil {
  return -1, raised
}
```

aisk commented on runtime/list_test.go:
```diff
@@ -91,6 +91,36 @@ func TestListCount(t *testing.T) {
+		{args: wrapArgs(newTestList(10, 20, 30), 20), want: NewInt(1).ToObject()},
+		{args: wrapArgs(newTestList(10, 20, 30), 20, newObject(intIndexType)), want: NewInt(1).ToObject()},
+		{args: wrapArgs(newTestList(0, 1, 2, 3, 4), 3, 3), want: NewInt(3).ToObject()},
+		{args: wrapArgs(newTestList(0, 2.0, 2, 3, 4, 2, 1, "foo"), 3, 3), want: NewInt(3).ToObject()},
```
@trotterdylan Hi, the repeated values test case is here.

aisk commented:
Updated.

trotterdylan commented on runtime/seq.go:
```diff
@@ -173,6 +173,26 @@ func seqFindFirst(f *Frame, iterable *Object, pred func(*Object) (bool, *BaseExc
+		var raised *BaseException
+		var eq *Object
+		var found bool
+		eq, raised = Eq(f, elem, o)
```
You can remove the eq and raised declarations above and then assign them this way:

```
eq, raised := Eq(f, elem, o)
```

Ditto for found, below.
